submariner/IPV6-datapath.md

```diff
@@ -0,0 +1,185 @@
+# Submariner Enhancement for IPV6 datapath
+
+<!-- Add link to issue/epic if available -->
+
+## Summary
+
+IPv4, IPV6 and dual-stack networking is supported for Kubernetes cluster starting in 1.21.
+IPV6 networking allowing assignment of IPv6 addresses.
+Dual-stack networking allowing the simultaneous assignment of both IPv4 and IPv6 addresses.
+
+IPv4/IPv6 dual-stack on your Kubernetes cluster provides the following features:
+
+* Dual-stack Pod networking (a single IPv4 and IPv6 address assignment per Pod)  
+* IPv4 and IPv6 enabled Services  
+* Pod off-cluster egress routing (eg. the Internet) via both IPv4 and IPv6 interfaces
+
+Currently, Submariner only supports IPV4 datapath and this proposal explains the changes required to support IPV6 and dual-stack clusters.
+
+## Overview
+
+Submariner’s control plane uses a central broker component to facilitate the exchange of metadata information between participating clusters.
+
+Both inter-cluster datapath and service-discovery (Lighthouse) apply their configuration based on the information received from control plane.
+
+The Broker, inter-cluster datapath and Lighthouse should be updated to support mixed clusters environment (e.g: dual-stack, V6 only , V4 only).
+
+### Control plane
+
+For mixed clusters environment we may experience connectivity issue.
+Some clusters and broker may not connect if the Broker’s cluster doesn’t support dual-stack.
+A dual-stack cluster should be selected as the broker in this case.
+
+### Inter-cluster datapath
+
+The inter-cluster datapath needs to support clusters with different networking configuration as described in the next table:
+
+| clusterA networking | clusterB networking | Supported connectivity type |
+| :---- | :---- | :---- |
+| V4 | V4 | V4 |
+| V4 | V6 | N/A |
+| V6 | dual-stack | V6 |
+| V6 | V6 | V6 |
+| V4 | dual-stack | V4 |
+| dual-stack | dual-stack | V4,V6 |
+
+### Service Discovery
+
+Lighthouse needs to handle imported services with a different IP family than the local cluster.
+This is necessary in a mixed cluster environment, as described below.
+
+|  | Local cluster networking | Imported service networking | Supported DNS record  |
+| :---- | :---- | :---- | :---- |
+| 1 | V4 | V4 | V4 |
+| 2 | V4 | V6 | N/A |
+| 3 | V6 | dual-stack | V6 |
+| 4 | V6 | V6 | V6 |
+| 5 | V4 | dual-stack | V4 |
+| 6 | dual-stack | dual-stack | V4,V6 |
+
+This can be achieved in one of the following ways:
+
+* Lighthouse ignores processing imported services IP addresses that don’t match local cluster networking configuration.
+* Lighthouse will process imported services regardless of local cluster networking configuration, and rely on local workloads dns requests.
+For example, in use case \#3 in the table above, Lighthouse DNS database will store both V4 and V6 addresses.
+However the local dns client will look for the V6 (AAAA) record.
```

Comment:
Normally, all the clients/apps make DNS requests for both v4/v6 and try connecting to them in parallel with a slight preference for v6. Please see Happy Eyeballs algo - https://datatracker.ietf.org/doc/html/rfc6555

Reply:
Is this case also for single-stack clients? 
I mean does V6 client/app try to connect to V4 address returned from DNS ?

Reply:
Yes, if the DNS server returns any v4 address, then the client will try to connect to it (which would fail if the server is not listening on it).

Reply:
Interesting, if V6 client tries to connect to IP V4 address I assume we should go with first approach

Reply:
If the local cluster supports only V4 and the remote cluster is v6 only (or dualStack), then returning V6 address as part of Lighthouse may not help since we do not use any intermediate proxies (or NAT mechanisms) to translate v4 to v6. So, approach 1 seems reasonable.

Reply:
The plan is to add [validation](https://github.com/submariner-io/enhancements/pull/240/files#diff-622b5a04b96a48ea8501651697d89c87465357f36b392c61e23d8f4d024ef9feR46) for clusters connectivity, 
so a cluster should join clusterset only if it can communicate with all existing clusters in the set.

Even after adding the validation we should still handle the following special cases:

1. client : single stack V4 , DNS : V4, V6 entries
2. client : single stack V6 , DNS : V4, V6 entries
3. client : dual-stack ,         DNS : V4 entry 
4. client : dual-stack ,         DNS : V6 entry 

Now, IIUC your comments, if we go with approach '2' things may fail for ^^ use cases.

just double checking, for use case#1, the V4 client may try to connect to the V6 address, right ?  (not sure how the IP header looks like in this case)